vernac/comInductive.ml

```diff
@@ -145,16 +126,78 @@ let pretype_ind_arity env sigma (loc, c, impls, pseudo_poly) =
     in
     sigma, (t, Retyping.relevance_of_sort s, concl, impls)
 
-let interp_cstrs env sigma impls mldata arity ind =
+let warn_deprecated_inductive_param_heuristic =
+  CWarnings.create ~name:"deprecated-inductive-param-heuristic" ~category:"deprecated"
+    (fun () -> Pp.(strbrk "The value inferred for some implicit arguments in " ++
+    strbrk"this declaration relies on a deprecated fragile heuristic. Please specify the " ++
+    strbrk"value of the argument or declare the inductive type as uniform using the " ++
+    strbrk"#[uniform] attribute."))
+
+let solve_uniform_inductive_params env evd indnames nparams =
+  let indnames = List.map (fun x -> GlobRef.VarRef x) indnames in
+  let has_resolved = ref false in
+  let evd = Evd.fold_undefined (fun evk ev_info evd' ->
+    match ev_info.Evd.evar_source with
+    | loc,Evar_kinds.ImplicitArg(r,(k,na),force) when List.mem_f GlobRef.equal r indnames ->
```

Comment:
Using the evar source seems more complicated than what I did, also IIUC it won't handle using explicit holes, eg
~~~
Inductive foo A := C : (_ : foo _).
~~~
I propose we go with my commit (https://github.com/coq/coq/commit/876b6f206748a3e4bcb26ee7b915b76653d9623e) instead, and don't deprecate as it doesn't seem especially fragile.

Reply:
Looking at the evar source makes sense only if the goal is to detect situations where the previous heuristic applied, which is what I did, to emit a warning only in these cases.

I think I prefer to deprecate it, as the cost/benefit ratio seems very low. The code for this inferrence is not simple, and the idea seem really ad-hoc (why would Coq make random bets at what the parameters value could be?).

But I guess I could be convinced by a real-world example where the hack is useful (i.e. where specifying uniform or the arguments is too painful). Do you have one?

Reply:
Uniform doesn't work if you have some non uniform parameters
Also IMO the cost is pretty low

Reply:
@What does your code do on `C : id foo _ -> foo`? Does it also try to fill up the hole?

Reply:
Yes

Reply:
No wait I didn't see te `id`, so no

Reply:
Where is reduction performed?

Reply:
So this is fragile. If you want to be robust, you'd have to share the code analyzing parameters in `indtypes.ml` (i.e. what positivity accepts). Not worth it, IMHO.

Reply:
I don't understand the meaning of fragile here.

Reply:
Take a definition where your heuristic applies. Modify it slighlty (add an `id`, a cast, a beta redex) and suddenly the behavior of inferrence changes. That's what I call fragile.

Reply:
Ha ha ha, we had the same discussion yesterday.
FTR, I gave up trying to convince @maximedenes that the approach it is not fragile... :-)

Reply:
Isn't the point of changing a definition to something convertible that it changes inference?

Reply:
I think we tend to introduce way too often some features that work "often". This is part of what makes Coq hard to use sometimes.

If we really want the heuristic, let's go ahead and do the same analysis as `indtypes.ml` (i.e. with reduction).